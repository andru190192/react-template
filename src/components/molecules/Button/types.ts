import { ButtonProps, IconButtonProps } from '@mui/material';

export interface IButtonProps extends ButtonProps {
  children: ButtonProps['children'];
  loading?: boolean;
}

export interface IIconButtonProps extends IconButtonProps {
  loading?: boolean;
  icon: IconButtonProps['children'];
}
