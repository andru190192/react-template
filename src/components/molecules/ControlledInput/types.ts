import { OutlinedTextFieldProps } from '@mui/material';
import { Control } from 'react-hook-form';

export interface ControlledInputProps extends OutlinedTextFieldProps {
  name: string;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  control: Control<any, any>;
}
