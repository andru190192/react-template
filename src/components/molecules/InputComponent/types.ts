import { OutlinedTextFieldProps } from '@mui/material';

export interface IInputComponentProps extends OutlinedTextFieldProps {
  helperText?: string;
  keepHelperTextSpace?: boolean;
}
