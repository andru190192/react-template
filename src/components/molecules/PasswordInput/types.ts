import { OutlinedTextFieldProps } from '@mui/material';
import { Control } from 'react-hook-form';

export interface PasswordInputProps extends OutlinedTextFieldProps {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  control: Control<any, any>;
}
