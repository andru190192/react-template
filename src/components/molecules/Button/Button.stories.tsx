import { Meta } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import Button from './Button';
import { IButtonProps } from './types';

const meta = {
  title: 'Components/Button',
  component: Button,
  tags: ['autodocs'],
  args: {
    variant:'contained',
    color:'primary',
    type:'submit',
    children: 'Text'
  },
} satisfies Meta<typeof Button>;

export default meta;

export const Example = ({  ...props }: IButtonProps) => {
  const handleClick = () => action(`is onClick`)();
  return (
      <Button {...props} onClick={handleClick} />
  );
};
