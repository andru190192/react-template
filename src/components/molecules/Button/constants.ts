export enum LOADING_SIZE {
  large = 24,
  medium = 20,
  small = 16,
}
