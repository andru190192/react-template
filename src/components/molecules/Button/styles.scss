.loading-button {
  position: relative;

  .loading {
    position: absolute;
  }

  .container-children {
    &.loading {
      position: relative;
      visibility: hidden;
    }
  }
}
