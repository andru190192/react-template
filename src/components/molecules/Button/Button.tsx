import {
  Box,
  Button as MUIButton,
  CircularProgress,
  IconButton,
} from '@mui/material';
import { IButtonProps, IIconButtonProps } from './types';

import { LOADING_SIZE } from './constants';
import './styles.scss';

export const Button = ({ children, loading, ...props }: IButtonProps) => {
  const { size } = props;

  return (
    <MUIButton className='loading-button' disabled={loading} {...props}>
      {loading && (
        <CircularProgress
          className='loading'
          size={LOADING_SIZE[size || 'small']}
          color='inherit'
        />
      )}

      <Box className={`container-children ${loading ? 'loading' : ''}`}>
        {children}
      </Box>
    </MUIButton>
  );
};

export default Button;


export const LoadingIconButton = ({
  loading,
  icon,
  ...props
}: IIconButtonProps) => {
  const { size } = props;

  return (
    <IconButton disabled={loading} {...props}>
      {loading ? (
        <CircularProgress
          size={LOADING_SIZE[size || 'small']}
          color='inherit'
        />
      ) : (
        icon
      )}
    </IconButton>
  );
};
