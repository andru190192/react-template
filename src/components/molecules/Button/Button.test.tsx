import { describe, expect, it, afterEach, vitest } from 'vitest';
import { cleanup, render } from '@testing-library/react';
import fireEvent from '@testing-library/user-event';
import Button from '.';
import { IButtonProps } from './types';

const clickMock = vitest.fn();

const props: IButtonProps = {
  variant:'contained',
  color:'primary',
  type:'submit',
  children: 'Log in',
} 

describe('<Button />', () => {
  afterEach(() => {
    cleanup();
  });

  it('Should render correctly', () => {
    const { container } = render(
      <Button {...props} /> 
    );

    expect(container).toMatchSnapshot();
  });

  it('Should dispatch click callback', async () => {
    const { getByText } = render(
      <Button {...props} onClick={clickMock} />
    );

    const loginBtn = getByText('Log in');

    await fireEvent.click(loginBtn);
    
    expect(loginBtn).toBeTruthy();
    expect(clickMock).toBeCalled();
  });
});
