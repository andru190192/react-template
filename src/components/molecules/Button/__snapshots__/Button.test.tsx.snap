// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Button /> > Should render correctly 1`] = `
<div>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium loading-button css-sghohy-MuiButtonBase-root-MuiButton-root"
    tabindex="0"
    type="submit"
  >
    <div
      class="container-children  MuiBox-root css-0"
    >
      Log in
    </div>
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</div>
`;
